File: android/app/src/main/java/com/locationtracker/core/api/APIError.kt
package com.locationtracker.core.api

import retrofit2.HttpException
import java.io.IOException

sealed class APIError : Exception() {
    object InvalidURL : APIError()
    object InvalidResponse : APIError()
    data class HttpStatus(val statusCode: Int) : APIError()
    data class DecodingFailed(val throwable: Throwable) : APIError()
    data class Underlying(val throwable: Throwable) : APIError()

    override val message: String?
        get() = when (this) {
            is InvalidURL -> "The API URL is invalid."
            is InvalidResponse -> "The server returned an invalid response."
            is HttpStatus -> "The server returned status code $statusCode."
            is DecodingFailed -> "Failed to decode the response: ${throwable.localizedMessage}"
            is Underlying -> throwable.localizedMessage
        }

    companion object {
        fun from(throwable: Throwable): APIError {
            return when (throwable) {
                is HttpException -> HttpStatus(throwable.code())
                is IOException -> Underlying(throwable)
                is APIError -> throwable
                else -> Underlying(throwable)
            }
        }
    }
}

File: ios/LocationTracker/Core/API/APIError.swift
import Foundation

enum APIError: LocalizedError {
  case invalidURL
  case invalidResponse
  case httpStatus(Int)
  case decodingFailed(Error)
  case underlying(Error)

  var errorDescription: String? {
    switch self {
    case .invalidURL:
      return "The API URL is invalid."
    case .invalidResponse:
      return "The server returned an invalid response."
    case .httpStatus(let statusCode):
      return "The server returned status code \(statusCode)."
    case .decodingFailed(let error):
      return "Failed to decode the response: \(error.localizedDescription)"
    case .underlying(let error):
      return error.localizedDescription
    }
  }
}
